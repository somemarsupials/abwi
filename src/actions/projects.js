import { api } from '../App';

export const actions = {
  FETCHING: 'FETCHING',
  FETCH_SUCCESS: 'FETCH_SUCCESS',
  FETCH_FAIL: 'FETCH_FAIL',
};

async function makeRequest() {
  return await fetch(`${api}/projects`);
};

export function projectsLoading(bool) {
  return {
    type: actions.FETCHING,
    isFetching: bool,
  };
};

export function projectsFetched(data) {
  return {
    type: actions.FETCH_SUCCESS,
    data: data,
  };
};

export function projectsNotFetched(error) {
  return {
    type: actions.FETCH_FAIL,
    error: error,
  };
};

export function fetchProjects() {
  return async (dispatch) => {
    let response, data, error;
    dispatch(projectsLoading(true));

    try {
      response = await makeRequest();
      data = await response.json();
    } 
    catch (e) {
      error = e.message;
    };

    dispatch(projectsLoading(false));

    if (!error && !response.ok) {
      error = response.ok;
    };

    if (error) {
      dispatch(projectsNotFetched(error));
    } else {
      dispatch(projectsFetched(data));
    };
  };
};
