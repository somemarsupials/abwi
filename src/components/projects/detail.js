import React from 'react';
import { ItemList } from '../items';
import { SectionBar } from '../common';

export default function(props) {
  return (
    <div>
      <SectionBar text={'Client'} />
      <span>{props.project.client.name}</span>
      <SectionBar text={'Items'} />
      <ItemList items={props.project.items}/>
    </div>
  );
};
