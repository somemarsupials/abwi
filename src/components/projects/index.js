export { default as ProjectPage } from './projectPage';
export { default as ProjectDetail } from './projectDetail';
