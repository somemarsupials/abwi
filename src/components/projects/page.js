import React from 'react';

import { TitleBar, InfoBar, ErrorBar, Button } from '../common/';
import Detail from './detail';

export default function(props) {
  let error = props.error || !props.project;
  let title = props.project && props.project.title || null;

  return (
    <div>
      <TitleBar text={title} /> 
      { error &&
        <ErrorBar text={'Could not load project'} detail={props.error} />
      }
      { !error && props.fetching && 
        <InfoBar text={'Fetching project...'} />
      }
      { !error && !props.fetching &&
        <Detail project={props.project} />
      }
    </div>
  );
};

