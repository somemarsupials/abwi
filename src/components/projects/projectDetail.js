import React from 'react';
import { SectionBar } from '../common';
import { ProjectsList } from '../allProjects';

export default function(props) {
  return (
    <div>
      <SectionBar text={'Client'} />
      <span>{props.project.client.name}</span>
      <SectionBar text={'Items'} />
      <ProjectsList items={props.project.items}/>
    </div>
  );
};
