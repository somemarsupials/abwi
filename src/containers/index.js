export { default as ProjectsContainer } from './projectsContainer';
export { default as ProjectContainer } from './projectContainer';
