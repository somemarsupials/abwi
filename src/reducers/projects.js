import { combineReducers } from 'redux';
import { actions } from '../actions/projects';
import { actions as modalActions } from '../actions/projectModal';

function fetching(state = false, action) {
  switch (action.type) {
    case actions.FETCH_LOADING:
      return action.active;
    default:
      return state;
  }
};

const initialState = {
  error: null,
  data: [],
};

function fetched(state = initialState, action) {
  switch (action.type) {
    case actions.FETCHED:
      return {
        error: action.error,
        data: action.data,
      };
    case modalActions.CREATED:
      return {
        error: state.error,
        data: state.data.concat([action.data]),
      };
    default:
      return state;
  }
};

export default combineReducers({
  fetching: fetching,
  response: fetched,
});
