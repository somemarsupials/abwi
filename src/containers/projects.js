import React, { Component } from 'react';
import { connect } from 'react-redux';

import { projectActions } from '../actions';
import { ProjectsOverview } from '../components/projects';


export class Projects extends Component {
  async componentDidMount() {
    let { api, success, fail } = this.props;
    this.props.fetch(api, success, fail);
  };

  render() {
    return (
      <ProjectsOverview
        projects={this.props.projects} 
        error={this.props.error}
      />
    );
  };
};


function mapStateToProps(state) {
  return { 
    projects: state.projects.data,
    error: state.projects.error,
  };
};


function mapDispatchToProps(dispatch) {
  let { fetchedProjects, projectFetchFailed } = projectActions;
  return {
    success: (projects) => {
      dispatch(fetchedProjects(projects))
    },
    fail: (error) => {
      dispatch(projectFetchFailed(error))
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Projects);
