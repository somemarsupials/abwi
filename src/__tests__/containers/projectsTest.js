import React from 'react';
import { shallow } from 'enzyme';

import { Projects } from '../../containers/projects';

describe('Project', () => {
  let props, projects, response, mockFetch; 

  beforeEach(() => {
    props = { fetchedProjects: jest.fn(), cannotFetch: jest.fn() };
  });

  describe('when mounted', () => {

    describe('when making API call', () => {
      beforeEach(() => {
        response = { ok: false };
        mockFetch = jest.fn().mockReturnValue(response);
        Object.assign(props, { api: 'api', fetch: mockFetch });
        projects = shallow(<Projects {...props} />);
      });

      it('gets correct resource', () => {
        expect(mockFetch).toBeCalledWith('api/project');
      });
    });

    describe('when API call succeeds', () => {
      beforeEach(() => {
        response = { ok: true, json: jest.fn().mockReturnValue('data') };
        mockFetch = jest.fn().mockReturnValue(response);
        Object.assign(props, { fetch: mockFetch });
        projects = shallow(<Projects {...props} />);
      });

      it('emits action with data', () => {
        expect(props.fetchedProjects).toBeCalledWith('data');
      });
    });

    describe('when API call throws error', () => {
      beforeEach(() => {
        mockFetch = jest.fn().mockImplementation(() => { 
          throw { message: 'error' }
        });
        Object.assign(props, { fetch: mockFetch });
        projects = shallow(<Projects {...props} />);
      });

      it('emits action with data', () => {
        expect(props.cannotFetch).toBeCalledWith('error');
      });
    });

    describe('when API call fails', () => {
      beforeEach(() => {
        response = { ok: false, status: 404 };
        mockFetch = jest.fn().mockReturnValue(response);
        Object.assign(props, { fetch: mockFetch });
        projects = shallow(<Projects {...props} />);
      });

      it('emits action with data', () => {
        let firstCall = props.cannotFetch.mock.calls[0];
        expect(firstCall.pop()).toMatch(/404/);
      });
    });
  });

  describe('when rendering', () => {
    let projectPage;

    beforeEach(() => {
      response = { ok: false, status: 404 };
      mockFetch = jest.fn().mockReturnValue(response);
      Object.assign(props, { 
        projects: [], 
        error: 'error', 
        fetch: mockFetch 
      });
      projectPage = shallow(<Projects {...props} />);
    });

    it('creates ProjectPage', () => {
      expect(projectPage).toMatchSnapshot();
    });
  });
});
