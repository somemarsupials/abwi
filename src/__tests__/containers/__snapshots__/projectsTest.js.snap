// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project when rendering creates ProjectPage 1`] = `
<Component
  error="error"
  projects={Array []}
/>
`;
